Add new functions: `pathListDir`, `pathIsFile`, `pathIsDirectory`

TracerDS opened the pull request:
Also refactored SharedUtil a little (`WIN32` might not be defined. `_WIN32` will always be defined)

Syntax: ~~`table fileListDir ( string path )`~~ `table pathListDir ( string path )`
returns table of entries on success, `nil` on failure

### Additional functions
~~`bool isFile ( string path )`~~ `bool pathIsFile ( string path )`
Returns true if path is a file, false if a directory or path doesn't exist

~~`bool isDirectory ( string path )`~~ `bool pathIsDirectory ( string path )`
Returns true if path is a directory, false if a file or path doesn't exist

### OOP functions
`path.listDir` => `pathListDir`
`path.isFile` => `pathIsFile`
`path.isDirectory` => `pathIsDirectory`

---

Resolves: #346

Lpsd committed:
Upgrade lunasvg from 2.3.5 to 2.3.8 (#3060)

patrikjuvonen committed:
Force rebuild

[ci skip]

patrikjuvonen committed:
Force rebuild

MTABot committed:
Visual Studio Update

VS2022: 17.6.33815.320 => 17.6.33829.357

[skip ci]

This is an automated commit to keep track of toolchain changes on the build server.
It applies to every MTA build after this commit until further notice.

MTABot committed:
Visual Studio Update

VS2019: 16.11.33801.447 => 16.11.33920.266

[skip ci]

This is an automated commit to keep track of toolchain changes on the build server.
It applies to every MTA build after this commit until further notice.

botder committed:
Bump build

botder committed:
Bump build

[ci skip]

botder committed:
Bump build

MTABot committed:
Visual Studio Update

VS2019: 16.11.33927.289 => 16.11.34031.81

[skip ci]

This is an automated commit to keep track of toolchain changes on the build server.
It applies to every MTA build after this commit until further notice.

MTABot committed:
Visual Studio Update

Build Tools 2022 (17.7.34031.279)

This is an automated commit to keep track of toolchain changes on the build server.
It applies to every MTA build after this commit until further notice. [skip ci]

TracerDS committed:
Add new function: `fileListDir`

Also refactored SharedUtil a little (`WIN32` might not be defined. `_WIN32` will always be defined)

TracerDS committed:
Fixed Client/Server code differences

Now `SharedUtil::ListDir` is crossplatform/crossservice

Pirulax commented:
Is this supposed to return a list of files in a directory?

TracerDS commented:
> Is this supposed to return a list of files in a directory?

Relative to a current resource, yes

botder commented:
Is the Unix code block functional with `if (!(dir = opendir("c:\\src\\")))`?
Also, don't you exclude the directory items in the `< C++17` block?
And last, how can the scripter differentiate between file and directory?

botder committed:
Bump build

[skip ci]

TracerDS committed:
Added `isFile`, `isDirectory`

Refactores `ListDir`, `FileExists` and `DirectoryExists` functions

TracerDS commented:
> Is the Unix code block functional with `if (!(dir = opendir("c:\\src\\")))`? Also, don't you exclude the directory items in the `< C++17` block? And last, how can the scripter differentiate between file and directory?

1. Changed path to the actual path. Should be functional now
2. Now including all files in a directory without adding the directory name before it too
3. via `isFile` and `isDirectory` functions (comment edited)

botder commented:
Wouldn't it make more sense to append a directory separator for directory items?
Checking each item whether it's a directory or file seems expensive?

TracerDS commented:
> Wouldn't it make more sense to append a directory separator for directory items?

It might cause some issues when messing with directories directly via that function (unless mta takes care of that already)

> Checking each item whether it's a directory or file seems expensive?

No, it's not really expensive. 

If MTA doesn't have problems with directories ending with `\` | `/` then sure, I can add a separator for directory entries

botder commented on Shared/sdk/SharedUtil.File.hpp:
```diff
@@ -1052,9 +1066,55 @@ int SharedUtil::File::Delete(const char* szFilename)
+    for (const auto& entry : fs::directory_iterator(szPath))
+    {
+        entries.push_back(entry.path().filename().string());
+    }
```
This is going to include non-directory and non-regular-file file types (block file, character file, socket file, symlinks) if I am not wrong. How about we only allow `entry.is_regular_file()` and `entry.is_directory()` here?

Also, you're using the throwing filesystem functions.

TracerDS replied:
> This is going to include non-directory and non-regular-file file types (block file, character file, socket file, symlinks) if I am not wrong. How about we only allow `entry.is_regular_file()` and `entry.is_directory()` here?
> 
> Also, you're using the throwing filesystem functions.

sure, we can do that. With throwing we can just use `try-catch` here

Pirulax commented:
I don't think `fileListDir` makes a lot of sense as a name.
Perhaps something like `getDirectoryEntries` or (or, I'd prefer to separate them, `getDirectoryFiles` and `getDirectoryFolders`, both with a `bool recursive` parameter as well.
Or, could even go with `pathGetFolders`, `pathGetFiles` [if we want to follow the `file` prefix convention thingy]

TracerDS commented:
> I don't think `fileListDir` makes a lot of sense as a name. Perhaps something like `getDirectoryEntries` or (or, I'd prefer to separate them, `getDirectoryFiles` and `getDirectoryFolders`, both with a `bool recursive` parameter as well. Or, could even go with `pathGetFolders`, `pathGetFiles` [if we want to follow the `file` prefix convention thingy]

its better to have one function that gets all entries that you can sort by using `isFile`/`isDirectory` functions.

TracerDS committed:
Updated due to review

Now only regular files and directories will be listed; marked `SharedUtil::ListDir` as `noexcept`

DNL291 commented:
There's the module FileSystem which already does the same and has some more features/functions btw.
Although could be added nativelly yes if the only purpose is to get files from a directory instead of using the module just for that.

TracerDS commented:
> There's the module FileSystem which already does the same and has some more features/functions btw.
> Although could be added nativelly yes if the only purpose is to get files from a directory instead of using the module just for that.

Using module for that will decrease performance (since you have to load the module, load its functions and events not to mention modules are very limited and you cant use almost any MTA functions there at all without refactoring the whole mta server).
Also that would only make the command work in server scripts only.

Implementing it directly in MTA solves all these issues

DNL291 commented:
> > There's the module FileSystem which already does the same and has some more features/functions btw.
> > Although could be added nativelly yes if the only purpose is to get files from a directory instead of using the module just for that.
> 
> Using module for that will decrease performance (since you have to load the module, load its functions and events not to mention modules are very limited and you cant use almost any MTA functions there at all without refactoring the whole mta server). Also that would only make the command work in server scripts only.
> 
> Implementing it directly in MTA solves all these issues

Uhm I'm not sure if we really have these issues in the (FileSystem) module. But I do agree that having the functions client-side and being directly implemented into MTA is better.

tederis commented:
> > There's the module FileSystem which already does the same and has some more features/functions btw.
> > Although could be added nativelly yes if the only purpose is to get files from a directory instead of using the module just for that.
> 
> Using module for that will decrease performance (since you have to load the module, load its functions and events not to mention modules are very limited and you cant use almost any MTA functions there at all without refactoring the whole mta server). Also that would only make the command work in server scripts only.
> 
> Implementing it directly in MTA solves all these issues

I doubt that modules can affect performance in a context of filesystem functions. The main disadvantage that modules have - they don't exist on a client side. So the only way to add proposed in this PR functions is to make them a part of MTA.

But I personally would not say that these functions are really needed on a client side. Typically, server(and therefore client scripts) know which files exist on a client and how to reach them, which makes sense from the security perspective.

Dutchman101 committed:
Force rebuild #3

For version control purposes

tederis commented on Shared/sdk/SharedUtil.File.hpp:
```diff
@@ -1052,9 +1066,60 @@ int SharedUtil::File::Delete(const char* szFilename)
+#else
+    #ifdef _WIN32
+    std::string search_path = szPath;
+    if (search_path.back() != '/')
```
When a string is empty method `back` has undefined behaviour. I would recommend to add a check `!search_path.empty()`.

tederis commented on Shared/sdk/SharedUtil.File.hpp:
```diff
@@ -53,14 +60,21 @@ bool SharedUtil::FileExists(const SString& strFilename)
-#ifdef WIN32
+#if _HAS_CXX17
+    namespace fs = std::filesystem;
+    return fs::is_directory(static_cast<const std::string&>(strPath));
```
Is static_cast really needed here?

TracerDS replied:
as below

tederis commented on Shared/sdk/SharedUtil.File.hpp:
```diff
@@ -29,22 +30,28 @@
-#ifdef WIN32
+#if _HAS_CXX17
+    namespace fs = std::filesystem;
+    return fs::is_regular_file(static_cast<const std::string&>(strFilename));
```
Is static_cast really needed here?

TracerDS replied:
to `std::string`? yes. Or we can use `.c_str()` but I think it would perform slower because it has to convert it to string first internally.

tederis replied:
This cannot be slower because the string copying will occur in any way. By using `c_str` you will achieve the same behaviour as you already have. `static_cast` is something that you should not use if you can.

tederis commented on Shared/sdk/SharedUtil.File.h:
```diff
@@ -102,6 +102,8 @@ namespace SharedUtil
     WString FromUTF8(const SString& strPath);
     SString ToUTF8(const WString& strPath);
 
+    std::vector<std::string> ListDir(const char* szPath) noexcept;
```
`noexcept` is dangerous here because `ListDir` method can throw an exception(bad_alloc for example).

TracerDS replied:
Bad_alloc can happen all the time, doesnt mean we should surround all mta code into multiple huge try-catch blocks. Other than that, no exceptions are thrown in ListDir function (except the possibility of an empty string but its easily solvable)

tederis replied:
`bad_alloc` is just an example of possible exceptions that can be thrown. `noexpect` in this case leads to `std::terminate` without even possibility to catch an exception somewhere upper in the stack. This is an example of so called ill-formed program.

tederis replied:
`noexcept` shall not be placed on a method that can even hypothetically throw an exception.

TracerDS replied:
Bad_alloc is the only exception here that can throw. All other are handled (fs functions throw under the fs_error class).
C functions dont throw at all. So there is nothing to worry about. If `bad_alloc` is thrown then mta cannot do anything anymore either way

tederis replied:
As I already said, `ill-formed, no diagnostic required`.

TracerDS committed:
Improved functionality

- Fixed some potential bugs,
- Now properly identifies C++17,
- Functions are now fully noexcept

botder committed:
Bump build [ci skip]

botder committed:
Bump build [ci skip]

Dutchman101 committed:
Force rebuild

Dutchman101 committed:
Force rebuild #2

For version control purposes

Dutchman101 committed:
Force rebuild #3

For version control purposes

Dutchman101 committed:
Force rebuild #4

For version control purposes, second-last one

TracerDS committed:
Added separate class for path-like functions

TracerDS committed:
Rebased branch

TheNormalnij commented on Shared/mods/deathmatch/logic/luadefs/CLuaPathDefs.cpp:
```diff
@@ -0,0 +1,93 @@
+
+    if (!DirectoryExists(strAbsPath))
+        return std::nullopt;
+
```
Warnings will be useful here

TracerDS replied:
What kind of warning here?

TheNormalnij replied:
"Unable to parse the path."
"Directory does not exist."
These warnings will be helpful during debugging and show potential issues.

TracerDS replied:
> "Unable to parse the path." "Directory does not exist." These warnings will be helpful during debugging and show potential issues.

sorry, but I've only done simple things with the parser. How would I throw a warning in ArgumentParser?

TheNormalnij replied:
Did you try `m_pScriptDebugging->LogWarning(luaVM, "message");` ?

TracerDS replied:
> Did you try `m_pScriptDebugging->LogWarning(luaVM, "message");` ?

`luaVM` is not accessible via ArgumentParser

TheNormalnij replied:
```
CLuaPathDefs::pathListDir(lua_State* luaVM, std::string path)
                                     ^^^^^^
 ```

TracerDS replied:
> ```
> CLuaPathDefs::pathListDir(lua_State* luaVM, std::string path)
>                                      ^^^^^^
> ```

oops, sorry my bad. I think Im too tired with it all 😄

TheNormalnij commented on Shared/mods/deathmatch/logic/luadefs/CLuaPathDefs.cpp:
```diff
@@ -0,0 +1,93 @@
+
+    CResource* pResource = pLuaMain->GetResource();
+    if (!CResourceManager::ParseResourcePathInput(path, pResource, &strAbsPath))
+        return false;
```
Warning

TheNormalnij commented on Shared/mods/deathmatch/logic/luadefs/CLuaPathDefs.cpp:
```diff
@@ -0,0 +1,93 @@
+
+    CResource* pResource = pLuaMain->GetResource();
+    if (!CResourceManager::ParseResourcePathInput(path, pResource, &strAbsPath))
+        return false;
```
Warning

TheNormalnij commented on Shared/sdk/SharedUtil.File.hpp:
```diff
@@ -1052,9 +1068,63 @@ int SharedUtil::File::Delete(const char* szFilename)
+                entries.push_back(entry.path().filename().string());
+        }
+    }
+    catch (...) {} // catch all possible errors and ignore them
+#else
+    #ifdef _WIN32
+    std::string search_path = szPath;
+    if (search_path.empty())
+        return {};
+
+    if (search_path.back() != '/')
+        search_path += "/*";
+
+    WIN32_FIND_DATA fd;
+    HANDLE          hFind = ::FindFirstFile(search_path.c_str(), &fd);
+    if (hFind == INVALID_HANDLE_VALUE)
+        return {};
+
+    do
+    {
+        entries.push_back(fd.cFileName);
+    } while (::FindNextFile(hFind, &fd));
+    ::FindClose(hFind);
+
+    #else
+    DIR*           dir;
+    struct dirent* ent;
+    if (!(dir = opendir(szPath)))
+        return {};
+
+    while ((ent = readdir(dir)))
+    {
+        entries.push_back(ent->d_name);
+    }
+    closedir(dir);
+    #endif
+#endif
+
+    return entries;
+}
```
```suggestion

#if __cplusplus >= 201703L
std::vector<std::string> SharedUtil::ListDir(const char* szPath) noexcept
{
    std::vector<std::string> entries;
    namespace fs = std::filesystem;
    if (!DirectoryExists(szPath))
        return {};

    try
    {
        for (const auto& entry : fs::directory_iterator(szPath))
        {
            if (entry.is_regular_file() || entry.is_directory())
                entries.push_back(entry.path().filename().string());
        }
    }
    catch (...) {} // catch all possible errors and ignore them

    return entries;
}
#endif
```
I think there is no reason to add dead code to the project.

TracerDS replied:
Some mta projects are still limited to c++14. Better safe than sorry

TheNormalnij replied:
But this code is unused and should be removed when we get C++ 17 there.
The codebase becomes more complex due to dead code. I would recommend keeping only the necessary code that can be tested.

TheNormalnij replied:
After the most recent force push, have you made these changes back?

TracerDS replied:
> After the most recent force push, have you made these changes back?

Nope. The code is used by `multiplayer_sa`. Cannot remove it.

Once `multiplayer_sa` gets converted to C++17+ we can remove old code

TheNormalnij commented on Shared/sdk/SharedUtil.File.hpp:
```diff
@@ -29,38 +30,53 @@
+    return s.st_mode & S_IFREG;
+    #endif
 #endif
 }
```
Can we use 'fs::is_regular_file' only here?

TheNormalnij commented on Shared/sdk/SharedUtil.File.hpp:
```diff
@@ -29,38 +30,53 @@
+    return s.st_mode & S_IFDIR;
+    #endif
 #endif
 }
```
Same question for  "fs::is_directory"

TheNormalnij commented on Shared/sdk/SharedUtil.File.h:
```diff
@@ -20,8 +20,8 @@ namespace SharedUtil
-    bool FileExists(const SString& strFilename);
-    bool DirectoryExists(const SString& strPath);
+    bool FileExists(const SString& strFilename) noexcept;
+    bool DirectoryExists(const SString& strPath) noexcept;
```
SString -> std::string

TracerDS replied:
Whats the point of SString if you dont use it anywhere?

SString inherits from std::string. Wouldnt make much difference

TheNormalnij replied:
We are trying to use only std::string in the new API. As far as I know, std::string -> SString casts have an allocation problem. So it's safe to use std::string here

TracerDS committed:
Merge branch 'master' of https://github.com/multitheftauto/mtasa-blue into 180923_Add-fileListDir

TracerDS committed:
Updated code

Cannot remove `#if __cplusplus >= 201703L` because client side code is using these functions (multiplayer_sa is c++14)

TracerDS committed:
Fixed Search & Replace issues

`WIN32_TESTING` was changed to `_WIN32_TESTING`. Changed back to the original definition

TracerDS commented on Client/mods/deathmatch/logic/CResource.h:
```diff
@@ -69,6 +69,7 @@ class CResource
     void                 SetResourceEntity(CClientEntity* pEntity) { m_pResourceEntity = pEntity; }
     class CClientEntity* GetResourceDynamicEntity() { return m_pResourceDynamicEntity; }
     void                 SetResourceDynamicEntity(CClientEntity* pEntity) { m_pResourceDynamicEntity = pEntity; }
+    SString              GetResourceDirectoryPath() { return GetResourceDirectoryPath(eAccessType::ACCESS_PUBLIC, ""); }
     SString              GetResourceDirectoryPath(eAccessType accessType, const SString& strMetaPath);
```
~~Oops, yeah forgot about that part.
Sure~~

Cant do that yet. Lets just merge these changes and SString will be taken care of in the next PR that will refactor these functions

TracerDS replied:
As far as I see all conflict have been resolved

TracerDS committed:
Merge branch 'master' into 180923_Add-fileListDir

TracerDS committed:
Merge branch 'master' into 180923_Add-fileListDir

Dutchman101 committed:
Merge branch 'master' into 180923_Add-fileListDir

Dutchman101 commented:
There's been many code reviews, so merging.
@TracerDS thanks, please add it to wiki though
